### ISSUE TITLE
Burrito loses focus when focusing on Gw2

### ISSUE THREAD

**Issue by Nuxmin** [AsherGlick/Burrito#25]
Hey there!

I could make it work like you stated using ArcDPS, however, another issue I've seen. Even if the Burrito is checked as always on top and I have gw on windowed. Every time you focus on gw the Burrito stays in the background.

I find no way to make it stay on top so it's only possible to see the path alternating with Alt Tab.

**Comment by AsherGlick** [comment]
Can you share what distro and window manager you are using?

**Comment by Nuxmin** [comment]
Sure!

I'm using Pop OS, it uses GNOME.

**Comment by AsherGlick** [comment]
Is guild wars in fullscreen or windowed-fullscreen?

**Comment by Nuxmin** [comment]
> Is guild wars in fullscreen or windowed-fullscreen?

Windowed.

**Comment by AsherGlick** [comment]
Unfortunately I cannot reproduce this issue on ubuntu using gnome and I do not have the resources to replicate a popos environment. If you or anyone else happen to figure it out, or are able to debug any more into the issue please add it here so we can work on a fix.

**Comment by coderedart** [comment]
i have the same issue. Endeavour/KDE latest packages. 

so, if you click on gw2 window, the whole overlay goes into background. but if you set the gw2 window as "keep below others" , overlay will now display over gw2 window. 
the issue with this work around is that you can't focus on burrito by clicking on its icon, you have to use alt-tab to actually focus input on burrito and then, you can click on burrito icon to get the main panel of "open marker file" and such. 

i just recommend sticking to window mode for now and maybe play gw2 with "borderless" decoration.

**Comment by coderedart** [comment]
i checked on gnome 40 and xfce 4.16 . the same issue with windowed full screen. with gnome, i couldn't even use windowed mode of gw2. 

According to distrowatch,
pop os has gnome 40.2 or 40.3, while arch/endeavour got gnome 40.3 . 
OTOH, ubuntu has the following gnome versions with the respective ubuntu releases. 
3.38.4 | 3.38.1 | 3.36.1
21.04  | 20.10  | 20.04 lts.

I *think* gnome 4 might have a bug that affects me and @Nuxmin . will need more people to report their distros/DE versions along with what issues they have with burrito to confirm. @AsherGlick what's your distro/gnome version? maybe you can put it in README as the main target that you develop for?

**Comment by coderedart** [comment]
https://github.com/AsherGlick/Burrito/issues/23#issuecomment-899065487
so, cinnamon works i guess.

**Comment by AsherGlick** [comment]
@coderedart I am using 20.04 LTS. I think you probably right about gnome 4 having some sort of bug, or weird interaction.

**Comment by coderedart** [comment]
> @coderedart I am using 20.04 LTS. I think you probably right about gnome 4 having some sort of bug, or weird interaction.

Think we should have a sort of table in readme with working combination or one primary target distro and wm. Cannot be bothered to support all variations of distros and wms. Maybe Ubuntu lts with gnome for starters

**Comment by coderedart** [comment]
I just installed ubuntu 20.04 lts, and the overlay was not working for me in gnome (3.38 version i think. i upgraded all packages to latest). i was researching this stuff for Jokolay, and i posted an issue on that repo, https://github.com/coderedart/jokolay/issues/1 . 

TLDR; 
Windowed fullscreen or fullscreen won't work reliably. unless we can use "Transient_For" hint, which needs us to dabble in x11.. 
Godot has a function to get the [raw x11 window handle ](https://docs.godotengine.org/en/stable/classes/class_os.html#class-os-method-get-native-handle) if we choose to use raw X11 commands via `OS.execute()`. we can also try other commands like https://docs.godotengine.org/en/stable/classes/class_os.html#class-os-method-move-window-to-foreground and call them every frame to put it above gw2 (don't know if it will work but worth a try). 
for now, i would just recommend telling people to use Window Mode and move burrito. and concentrate on getting gw2 window size via burrito_link.

**Comment by royalmustard** [comment]
>we can also try other commands like https://docs.godotengine.org/en/stable/classes/class_os.html#class-os-method-move-window-to-foreground and call them every frame to put it above gw2 (don't know if it will work but worth a try).

Tried that, it did not work

**Comment by royalmustard** [comment]
I found out why Burrito was eating the input when setting WM_TRANSIENT_FOR: It was because I used my own build which was still calling `OS.move_window_to_foreground()` in godot every frame. So mouse passthrough worked but because burrito was always above gw2, it got the keyboard. WM_TRANSIENT_FOR almost works, however burrito still receives no mouse inputs.

**Comment by AsherGlick** [comment]
This is great! I look forward to seeing the pr. Does the mouse capture region work properly when you are tabbed into burrito or do you have to tab back into Guildwars when you are done?

**Comment by royalmustard** [comment]
I guess we can close this now?

**Comment by AsherGlick** [comment]
We can close this issue. If users are still experiencing similar problems we should open a new one referencing this one.

**Comment by coderedart** [comment]
I get the following error
```
Godot Engine v3.3.3.stable.official.b973f997f - https://godotengine.org
WARNING: initialize: Error setting locale modifiers
   At: platform/x11/os_x11.cpp:153.
WARNING: initialize: XOpenIM failed
   At: platform/x11/os_x11.cpp:200.
WARNING: initialize: XCreateIC couldn't create xic
   At: platform/x11/os_x11.cpp:510.
OpenGL ES 3.0 Renderer: NVIDIA GeForce GTX 1070 Ti/PCIe/SSE2
OpenGL ES Batching: ON
 
pre
post init
Setting transient for
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: FailedToConnect', src/lib.rs:34:61
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
ERROR: <native>: gdnative-core: method panicked (check stderr for output)
   At: src/lib.rs:32.
```
and then burrito goes on to run normally. latest kde/arch. 
any requirements on how i should be starting gw2 or something? its a freshly installed gw2 from lutris with the lutris-gw2-6.14 runtime.

**Comment by AsherGlick** [comment]
This seems to be panicking as a result of failing to connect to the default x11 server
https://github.com/AsherGlick/Burrito/blob/fb2ef57e73f8c41e453c62dbc3b843cdd383d1b0/burrito-fg/src/lib.rs#L34

https://docs.rs/breadx/1.2.1/breadx/display/type.DisplayConnection.html says it can return errors from `from_connection`

https://docs.rs/breadx/1.2.1/breadx/display/struct.BasicDisplay.html#method.from_connection says this error is due to the x server rejecting the connection.

@royalmustard any ideas?

@coderedart when you say goes on to run normally, do you mean that it is still effected by this issue where burrito goes behind the gw2 window when gw2 is focused?

**Comment by coderedart** [comment]
aye. it goes behind in windowed fullscreen, and when i bring it forward with alt-tab, its just a black screen.
in windowed mode, its fine. but maybe its just my system, someone with not arch/kde can also try it. I will go see if it works on mx/xfce

**Comment by coderedart** [comment]
godot on mxlinux can't build burrito for some reason. so, i just tried arch/gnome and i get the same crash, but this time, i can't even see the burrito icon.  i will just see what @royalmustard tested this on and try that.

**Comment by royalmustard** [comment]
> i bring it forward with alt-tab, its just a black screen.

This seems to be an issue with the compositor not working. I had the same issue with Manjaro/kde and the kwin compositor. Try turning off compositing in kwin and install picom (or another compositor), that is what worked for me.

As for the crash, are you running pure X or Xwayland? I tested it on pure X on aforementioned setup (Manjaro, kwin, picom).
Godot also seems to have some problems with X (as indicated in your output) which do not appear when I run it.

**Comment by coderedart** [comment]
@royalmustard pure X. Just normal kde installation.i have nvidia card, so Wayland is a not an option. Maybe we should get someone else to check it. Maybe your system has a library installed that I don't?

**Comment by royalmustard** [comment]
which drivers are you running?

**Comment by coderedart** [comment]
@royalmustard nvidia proprietary latest

**Comment by royalmustard** [comment]
The alt tabbing is the only way we can get it to work on windowed fullscreen. Its a feature, not a bug

**Comment by royalmustard** [comment]
Since it seems to work on Ubuntu and not on Arch, maybe compare your Xorg installations (version, Xorg.conf, addons, etc.). Maybe share your arch findings so others can try to reproduce the bug.

**Comment by coderedart** [comment]
i can't even get breadx example to run. it fails to connect for some reason.

**Comment by royalmustard** [comment]
Try if you can get some xcb or xlib stuff to run, so we can see if it is a problem with breadx.

**Comment by coderedart** [comment]
everything else runs. it was fine when i was using x11rb to try the _NET_WINDOW_TYPE flags thingy. it seems breadx cant use xauth yet? https://github.com/bread-graphics/breadx/issues/8

**Comment by royalmustard** [comment]
I don't quite understand the byte parsing code going on in the breadx authorisation functions, but failing to authorize should give a different error than failing to connect (see bread-graphics/breadx/pull/10). According to the X protocol, on refusing the connection (which is something different than not being authorized), X should send a reason in the connection response. (However I don't quite understand enough of it to add functionality to read it). Maybe there is something in Xorg.log, but I am not sure.

On another note, I got someone to test it on arch with i3 and picom, where it seemed to run fine.

**Comment by AsherGlick** [comment]
This is really great progress. I think this has been scoped down enough to be its own issue at this point so it can be better tracked. @coderedart or @royalmustard can you transition this into a new issue with any system details or breadx code snippets that are necessary?
